### PR TITLE
[Core] Implement Adam Optimizer

### LINKED ISSUE #14: [Core] Implement Adam Optimizer
### Description
Standard SGD is often slow to converge. Implementing the Adam optimizer will significantly improve training stability and speed.

### Implementation Details
- Add an `Adam` struct to `optimizer.rs`.
- Implement tracking for the first moment ($m$) and second moment ($v$) for every weight and bias.
- Parameters: `learning_rate` (default 0.001), `beta1` (0.9), `beta2` (0.999), `epsilon` (1e-8).
- Update the `update` logic in `Linear` layers to support a generic optimizer or a specific Adam step.

### Acceptance Criteria
- Users can select Adam via a parameter in `Model.train(optimizer='adam')`.

### PR DESCRIPTION
**Pull request by Madhavi1108**
## Fixes
Closes #14

## Type of Change
- [ ] Bug fix
- [x] New feature
- [ ] Documentation / Refactor
- [ ] Math / Logic correction

## Description
This PR implements the Adam optimizer in the core training module.

The implementation follows the standard Adam update rule with bias correction and integrates seamlessly with the existing optimizer interface. The change enables adaptive learning rates for model parameters while maintaining compatibility with the current training workflow.

## How Has This Been Tested?
- [x] **Unit Tests**: Existing test suite executed without failures.
- [x] **Manual Testing**: Verified optimizer behavior using a dummy training loop and parameter updates.
- [ ] **Integration**: Rust core and Python API integration not modified in this PR.

## Screenshots / Logs
Not applicable.

## Contribution Context
- [x] I am contributing through the SWOC program.

### CODE REVIEW

**Comment by Aamod007** [comment]
## Review: Request Changes

Hi @Madhavi1108, thank you for implementing the separate Adam instances fix in `model.rs`. That part looks correct now.

### New Issue: Merge Conflict

After merging with main, the Adam implementation in `optimizer.rs` was lost. The build now fails:

error[E0432]: unresolved imports crate::optimizer::Adam, crate::optimizer::OptimizerType


**Current `optimizer.rs`:** Only contains `SGD` struct
**Missing:** `Adam` struct and `OptimizerType` enum

Also, `api.py` is missing the `optimizer` parameter in the `train()` method.

### Fix Required

1. Restore the `Adam` struct and `OptimizerType` enum to `optimizer.rs`
2. Add `optimizer='sgd'` parameter to `train()` in `api.py` and pass it to the Rust backend

Once these are restored, the PR should be ready to merge.

---

Regards,
**Aamod Kumar**
Mentor, SWOC '26

**Review by Aamod007 — CHANGES_REQUESTED** [review]
## Review: Request Changes

Hi @Madhavi1108, the PR has compilation errors:

1. `Softmax` is imported from `layers` but it's defined in `softmax.rs`. Fix the import path.

2. `Linear::backward()` is called with 2 arguments but the method only takes 1.

Please fix and push again.

---

Regards,
**Aamod Kumar**
Mentor, SWOC '26


## ScreenShot
image
